### ISSUE TITLE
[API] Tensor labels by strings; contraction based on labels

### ISSUE THREAD

**Issue by chiamin** [kaihsin/Cytnx#83]
I still recommend to use strings for tensors' labels, and make an option to use the labels to do the contraction. The design is to specify which label in tensor `A` is to be contracted with which label in tensor `B`. The idea is to **not** remember the order of indices when doing a contraction. The reasons for this suggestion are
1. Labels are easier to read and remember than the order of the indices. When I worked with Uli's code, which uses order, I needed to constantly ask the developer the correct order in every 1 month.
2. The order might change after a contraction, but the labels remain the same.
3. The readability is important when building a big program and work with other people's code.

Basically the situation I am thinking is that, imaging you are using an algorithm that generate an MPS for you, and you will use this MPS to do time evolution. It is difficult to trace the order of the indices by reading the code, but it could be more readable if |all the tensors have index labels like "left", "right", "top".

This design basically follows ITensor's logic that the users should not worry about the order of the indices, but with less restriction and nice flexibility. For example in ITensor it is annoying to do a simple contraction of two same tensors like

--A--A--

which we can do it easily.

**Comment by kaihsin** [comment]
All the labels are now changed to string after v0.9
